File: cpp/src/dense_ladder_order_book.cpp
#include "llab/dense_ladder_order_book.hpp"

#include <algorithm>
#include <limits>
#include <stdexcept>

namespace llab::dense_ladder_order_book {
namespace {
std::uint64_t mix(std::uint64_t x) {
    x += 0x9E3779B97F4A7C15ULL;
    x = (x ^ (x >> 30U)) * 0xBF58476D1CE4E5B9ULL;
    x = (x ^ (x >> 27U)) * 0x94D049BB133111EBULL;
    return x ^ (x >> 31U);
}
std::size_t capacity(std::size_t n) {
    std::size_t c = 8, target = std::max<std::size_t>(8, n * 2);
    while (c < target)
        c <<= 1U;
    return c;
}
} // namespace

OrderBook::OrderBook(std::size_t maximum_live_orders, Price minimum_price, std::size_t price_count)
    : slots_(capacity(maximum_live_orders)), bids_(price_count), asks_(price_count),
      minimum_price_(minimum_price) {
    if (price_count == 0 || minimum_price > std::numeric_limits<Price>::max() - (price_count - 1))
        throw std::invalid_argument("invalid ladder range");
}
void OrderBook::apply(const Event &e) {
    switch (e.type) {
    case EventType::Add:
        add_order(e.order_id, e.side, e.price_ticks, e.quantity);
        return;
    case EventType::Cancel:
        reduce(e.order_id, e.quantity, false);
        return;
    case EventType::Execute:
        reduce(e.order_id, e.quantity, true);
        return;
    case EventType::Delete:
    case EventType::OrderDelete:
        remove_order(e.order_id);
        return;
    case EventType::Replace:
        if (e.order_id == e.replacement_order_id)
            throw std::invalid_argument("replace id");
        remove_order(e.order_id);
        add_order(e.replacement_order_id, e.side, e.price_ticks, e.quantity);
        return;
    case EventType::OrderUpsert:
        upsert(e.order_id, e.side, e.price_ticks, e.quantity);
        return;
    }
}
std::optional<Level> OrderBook::best_bid() const {
    return best_bid_ ? std::optional<Level>{{*best_bid_, bids_[offset(*best_bid_)]}} : std::nullopt;
}
std::optional<Level> OrderBook::best_ask() const {
    return best_ask_ ? std::optional<Level>{{*best_ask_, asks_[offset(*best_ask_)]}} : std::nullopt;
}
std::size_t OrderBook::live_order_count() const noexcept {
    return live_;
}
std::size_t OrderBook::order_index_capacity() const noexcept {
    return slots_.size();
}
std::uint64_t OrderBook::state_digest() const {
    std::uint64_t h = 14695981039346656037ULL;
    auto b = [&](std::uint8_t x) { h = (h ^ x) * 1099511628211ULL; };
    auto u = [&](std::uint64_t x) {
        for (unsigned s = 0; s < 64; s += 8)
            b(static_cast<std::uint8_t>(x >> s));
    };
    u(live_);
    for (const auto *l : {&bids_, &asks_}) {
        std::size_t n = 0;
        for (auto q : *l)
            if (q)
                ++n;
        u(n);
        for (std::size_t i = 0; i < l->size(); ++i)
            if ((*l)[i]) {
                u(minimum_price_ + i);
                u((*l)[i]);
            }
    }
    std::vector<OrderId> ids;
    for (const auto &s : slots_)
        if (s.state == State::Occupied)
            ids.push_back(s.id);
    std::sort(ids.begin(), ids.end());
    for (auto id : ids) {
        const auto &s = slots_[find(id)];
        u(id);
        b(static_cast<std::uint8_t>(s.order.side));
        u(s.order.price);
        u(s.order.quantity);
    }
    return h;
}
std::size_t OrderBook::find(OrderId id) const {
    const auto mask = slots_.size() - 1;
    for (std::size_t i = mix(id) & mask, n = 0; n < slots_.size(); ++n, i = (i + 1) & mask) {
        const auto &s = slots_[i];
        if (s.state == State::Empty)
            return slots_.size();
        if (s.state == State::Occupied && s.id == id)
            return i;
    }
    return slots_.size();
}
std::size_t OrderBook::insert_slot(OrderId id) const {
    const auto mask = slots_.size() - 1;
    auto tomb = slots_.size();
    for (std::size_t i = mix(id) & mask, n = 0; n < slots_.size(); ++n, i = (i + 1) & mask) {
        const auto &s = slots_[i];
        if (s.state == State::Empty)
            return tomb == slots_.size() ? i : tomb;
        if (s.state == State::Tombstone && tomb == slots_.size())
            tomb = i;
        if (s.state == State::Occupied && s.id == id)
            return slots_.size();
    }
    return tomb;
}
OrderBook::Slot &OrderBook::require(OrderId id) {
    auto i = find(id);
    if (i == slots_.size())
        throw std::invalid_argument("unknown id");
    return slots_[i];
}
std::size_t OrderBook::offset(Price price) const {
    if (price < minimum_price_ || price - minimum_price_ >= bids_.size())
        throw std::invalid_argument("price outside ladder");
    return static_cast<std::size_t>(price - minimum_price_);
}
std::vector<Quantity> &OrderBook::ladder(Side s) {
    return s == Side::Bid ? bids_ : asks_;
}
const std::vector<Quantity> &OrderBook::ladder(Side s) const {
    return s == Side::Bid ? bids_ : asks_;
}
void OrderBook::add_order(OrderId id, Side side, Price price, Quantity q) {
    if (!id || !q)
        throw std::invalid_argument("add");
    auto i = insert_slot(id);
    if (i == slots_.size())
        throw std::invalid_argument("index");
    add_level(side, price, q);
    slots_[i] = {id, {side, price, q}, State::Occupied};
    ++live_;
}
void OrderBook::remove_order(OrderId id) {
    auto &s = require(id);
    subtract_level(s.order.side, s.order.price, s.order.quantity);
    s.state = State::Tombstone;
    --live_;
}
void OrderBook::reduce(OrderId id, Quantity q, bool full) {
    auto &s = require(id);
    if (!q || q > s.order.quantity || (!full && q == s.order.quantity))
        throw std::invalid_argument("reduce");
    subtract_level(s.order.side, s.order.price, q);
    if (q == s.order.quantity) {
        s.state = State::Tombstone;
        --live_;
    } else
        s.order.quantity -= q;
}
void OrderBook::upsert(OrderId id, Side side, Price price, Quantity q) {
    if (!id || !q)
        throw std::invalid_argument("upsert");
    auto i = find(id);
    if (i == slots_.size()) {
        add_order(id, side, price, q);
        return;
    }
    auto &s = slots_[i];
    subtract_level(s.order.side, s.order.price, s.order.quantity);
    add_level(side, price, q);
    s.order = {side, price, q};
}
void OrderBook::add_level(Side side, Price price, Quantity q) {
    auto &l = ladder(side);
    auto &i = l[offset(price)];
    if (std::numeric_limits<Quantity>::max() - i < q)
        throw std::overflow_error("overflow");
    i += q;
    if (side == Side::Bid) {
        if (!best_bid_ || price > *best_bid_)
            best_bid_ = price;
    } else if (!best_ask_ || price < *best_ask_)
        best_ask_ = price;
}
void OrderBook::subtract_level(Side side, Price price, Quantity q) {
    auto &l = ladder(side);
    auto &i = l[offset(price)];
    if (!q || q > i)
        throw std::logic_error("level");
    i -= q;
    if (i == 0 && ((side == Side::Bid && best_bid_ && price == *best_bid_) ||
                   (side == Side::Ask && best_ask_ && price == *best_ask_)))
        refresh_best(side);
}
void OrderBook::refresh_best(Side side) {
    const auto &l = ladder(side);
    if (side == Side::Bid) {
        best_bid_.reset();
        for (std::size_t i = l.size(); i-- > 0;)
            if (l[i]) {
                best_bid_ = minimum_price_ + i;
                break;
            }
    } else {
        best_ask_.reset();
        for (std::size_t i = 0; i < l.size(); ++i)
            if (l[i]) {
                best_ask_ = minimum_price_ + i;
                break;
            }
    }
}
} // namespace llab::dense_ladder_order_book

File: rust/src/dense_ladder_order_book.rs
use crate::parity_order_book::{Event, EventType, Level, OrderId, Price, Quantity, Side};

#[derive(Clone, Copy, PartialEq)]
enum State {
    Empty,
    Occupied,
    Tombstone,
}
#[derive(Clone, Copy)]
struct Order {
    side: Side,
    price: Price,
    quantity: Quantity,
}
#[derive(Clone, Copy)]
struct Slot {
    id: OrderId,
    order: Order,
    state: State,
}
impl Default for Slot {
    fn default() -> Self {
        Self { id: 0, order: Order { side: Side::Bid, price: 0, quantity: 0 }, state: State::Empty }
    }
}
pub struct OrderBook {
    slots: Vec<Slot>,
    bids: Vec<Quantity>,
    asks: Vec<Quantity>,
    minimum: Price,
    best_bid: Option<Price>,
    best_ask: Option<Price>,
    live: usize,
}
impl OrderBook {
    pub fn new(max: usize, minimum: Price, count: usize) -> Self {
        assert!(count > 0 && minimum <= u64::MAX - (count as u64 - 1));
        Self {
            slots: vec![Slot::default(); capacity(max)],
            bids: vec![0; count],
            asks: vec![0; count],
            minimum,
            best_bid: None,
            best_ask: None,
            live: 0,
        }
    }
    pub fn apply(&mut self, e: Event) -> Result<(), &'static str> {
        match e.event_type {
            EventType::Add => self.add(e.order_id, e.side, e.price_ticks, e.quantity),
            EventType::Cancel => self.reduce(e.order_id, e.quantity, false),
            EventType::Execute => self.reduce(e.order_id, e.quantity, true),
            EventType::Delete | EventType::OrderDelete => self.remove(e.order_id),
            EventType::Replace => {
                if e.order_id == e.replacement_order_id {
                    return Err("replace id");
                }
                self.remove(e.order_id)?;
                self.add(e.replacement_order_id, e.side, e.price_ticks, e.quantity)
            }
            EventType::OrderUpsert => self.upsert(e.order_id, e.side, e.price_ticks, e.quantity),
        }
    }
    pub fn best_bid(&self) -> Option<Level> {
        self.best_bid
            .map(|p| Level { price_ticks: p, quantity: self.bids[self.offset(p).unwrap()] })
    }
    pub fn best_ask(&self) -> Option<Level> {
        self.best_ask
            .map(|p| Level { price_ticks: p, quantity: self.asks[self.offset(p).unwrap()] })
    }
    pub fn live_order_count(&self) -> usize {
        self.live
    }
    pub fn order_index_capacity(&self) -> usize {
        self.slots.len()
    }
    pub fn state_digest(&self) -> u64 {
        let mut h = 14_695_981_039_346_656_037u64;
        hash_u(&mut h, self.live as u64);
        for l in [&self.bids, &self.asks] {
            hash_u(&mut h, l.iter().filter(|&&q| q != 0).count() as u64);
            for (i, &q) in l.iter().enumerate() {
                if q != 0 {
                    hash_u(&mut h, self.minimum + i as u64);
                    hash_u(&mut h, q)
                }
            }
        }
        let mut ids: Vec<_> =
            self.slots.iter().filter(|s| s.state == State::Occupied).map(|s| s.id).collect();
        ids.sort_unstable();
        for id in ids {
            let s = self.slots[self.find(id).unwrap()];
            hash_u(&mut h, id);
            h = (h ^ (s.order.side as u8 as u64)).wrapping_mul(1_099_511_628_211);
            hash_u(&mut h, s.order.price);
            hash_u(&mut h, s.order.quantity)
        }
        h
    }
    fn find(&self, id: OrderId) -> Option<usize> {
        let m = self.slots.len() - 1;
        let mut i = mix(id) as usize & m;
        for _ in 0..self.slots.len() {
            let s = self.slots[i];
            if s.state == State::Empty {
                return None;
            }
            if s.state == State::Occupied && s.id == id {
                return Some(i);
            }
            i = (i + 1) & m
        }
        None
    }
    fn insertion(&self, id: OrderId) -> Result<usize, &'static str> {
        let m = self.slots.len() - 1;
        let (mut i, mut tomb) = (mix(id) as usize & m, None);
        for _ in 0..self.slots.len() {
            let s = self.slots[i];
            if s.state == State::Empty {
                return Ok(tomb.unwrap_or(i));
            }
            if s.state == State::Tombstone && tomb.is_none() {
                tomb = Some(i)
            }
            if s.state == State::Occupied && s.id == id {
                return Err("index");
            }
            i = (i + 1) & m
        }
        tomb.ok_or("index")
    }
    fn offset(&self, p: Price) -> Result<usize, &'static str> {
        if p < self.minimum {
            return Err("price");
        }
        let i = (p - self.minimum) as usize;
        if i >= self.bids.len() { Err("price") } else { Ok(i) }
    }
    fn add(&mut self, id: OrderId, side: Side, p: Price, q: Quantity) -> Result<(), &'static str> {
        if id == 0 || q == 0 {
            return Err("add");
        }
        let i = self.insertion(id)?;
        self.add_level(side, p, q)?;
        self.slots[i] =
            Slot { id, order: Order { side, price: p, quantity: q }, state: State::Occupied };
        self.live += 1;
        Ok(())
    }
    fn remove(&mut self, id: OrderId) -> Result<(), &'static str> {
        let i = self.find(id).ok_or("id")?;
        let o = self.slots[i].order;
        self.sub_level(o.side, o.price, o.quantity)?;
        self.slots[i].state = State::Tombstone;
        self.live -= 1;
        Ok(())
    }
    fn reduce(&mut self, id: OrderId, q: Quantity, full: bool) -> Result<(), &'static str> {
        let i = self.find(id).ok_or("id")?;
        let o = self.slots[i].order;
        if q == 0 || q > o.quantity || (!full && q == o.quantity) {
            return Err("reduce");
        }
        self.sub_level(o.side, o.price, q)?;
        if q == o.quantity {
            self.slots[i].state = State::Tombstone;
            self.live -= 1
        } else {
            self.slots[i].order.quantity -= q
        }
        Ok(())
    }
    fn upsert(
        &mut self,
        id: OrderId,
        side: Side,
        p: Price,
        q: Quantity,
    ) -> Result<(), &'static str> {
        if id == 0 || q == 0 {
            return Err("upsert");
        }
        if let Some(i) = self.find(id) {
            let o = self.slots[i].order;
            self.sub_level(o.side, o.price, o.quantity)?;
            self.add_level(side, p, q)?;
            self.slots[i].order = Order { side, price: p, quantity: q };
            Ok(())
        } else {
            self.add(id, side, p, q)
        }
    }
    fn add_level(&mut self, side: Side, p: Price, q: Quantity) -> Result<(), &'static str> {
        let i = self.offset(p)?;
        let l = if side == Side::Bid { &mut self.bids } else { &mut self.asks };
        l[i] = l[i].checked_add(q).ok_or("overflow")?;
        if side == Side::Bid {
            if self.best_bid.map_or(true, |x| p > x) {
                self.best_bid = Some(p)
            }
        } else if self.best_ask.map_or(true, |x| p < x) {
            self.best_ask = Some(p)
        }
        Ok(())
    }
    fn sub_level(&mut self, side: Side, p: Price, q: Quantity) -> Result<(), &'static str> {
        let i = self.offset(p)?;
        let l = if side == Side::Bid { &mut self.bids } else { &mut self.asks };
        if q == 0 || q > l[i] {
            return Err("level");
        }
        l[i] -= q;
        if l[i] == 0
            && ((side == Side::Bid && self.best_bid == Some(p))
                || (side == Side::Ask && self.best_ask == Some(p)))
        {
            self.refresh(side)
        }
        Ok(())
    }
    fn refresh(&mut self, side: Side) {
        let l = if side == Side::Bid { &self.bids } else { &self.asks };
        let found = if side == Side::Bid {
            l.iter().rposition(|&q| q != 0)
        } else {
            l.iter().position(|&q| q != 0)
        };
        if side == Side::Bid {
            self.best_bid = found.map(|i| self.minimum + i as u64)
        } else {
            self.best_ask = found.map(|i| self.minimum + i as u64)
        }
    }
}
fn capacity(n: usize) -> usize {
    (n.saturating_mul(2)).max(8).next_power_of_two()
}
fn mix(mut x: u64) -> u64 {
    x = x.wrapping_add(0x9E37_79B9_7F4A_7C15);
    x = (x ^ (x >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    x = (x ^ (x >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    x ^ (x >> 31)
}
fn hash_u(h: &mut u64, x: u64) {
    for s in (0..64).step_by(8) {
        *h = (*h ^ ((x >> s) as u8 as u64)).wrapping_mul(1_099_511_628_211)
    }
}
